dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -103,68 +125,119 @@ data SourceCodeFragment =
 
-            sourceCodeFragment = SourceCodeFragment varSrc (VariableDeclaration varDecl)
-
-            fromExpr' = go (Context.insert variable varDecl context) expr'
+            let varSrc = makeSrcForLabel srcEnd0 srcStart1 variable
+            let varDecl = VarDecl varSrc variable bindingDhallType
 
+            Writer.tell [SourceCodeFragment varSrc (VariableDeclaration varDecl)]
+            infer (Context.insert variable varDecl context) expr'
 
         Note src (Var (V name index)) ->
             case Context.lookup name index context of
-                Nothing -> []
-                Just varDecl -> [SourceCodeFragment src $ VariableUse varDecl]
+                Nothing -> return AnyType
+                Just varDecl@(VarDecl _ _ t) -> do
+                    Writer.tell [SourceCodeFragment src $ VariableUse varDecl]
+                    return t
 
         Lam (FunctionBinding
                 (Just Src{srcEnd = srcEnd0})
                 variable
                 (Just Src{srcStart = srcStart1})
                 _
-                t) expr ->
-            sourceCodeFragment : fromAnnotation ++ fromExpr
-          where
-            varSrc = makeSrcForLabel srcEnd0 srcStart1 variable
-            varDecl = VarDecl varSrc variable
-            sourceCodeFragment = SourceCodeFragment varSrc (VariableDeclaration varDecl)
+                t) expr -> do
+            dhallType <- infer context t
 
-            fromAnnotation = go context t
-            fromExpr = go (Context.insert variable varDecl context) expr
-        e -> concatMap (go context) $ Lens.toListOf Core.subExpressions e
+            let varSrc = makeSrcForLabel srcEnd0 srcStart1 variable
+            let varDecl = VarDecl varSrc variable dhallType
 
-    makeSrcForLabel srcStart srcEnd variable = Src {..}
-      where
-        realLength = getSourceColumn srcEnd - getSourceColumn srcStart
-        srcText =
-            if Text.length variable == realLength then variable
-            else "`" <> variable <> "`"
+            Writer.tell [SourceCodeFragment varSrc (VariableDeclaration varDecl)]
+            infer (Context.insert variable varDecl context) expr
+
+        Field e (FieldSelection (Just Src{srcEnd=posStart}) label (Just Src{srcStart=posEnd})) -> do
+            fields <- do
+                dhallType <- infer context e
+                case dhallType of
+                    AnyType -> return mempty
+                    RecordLiteral s -> return $ Set.toList s
+
+            let src = makeSrcForLabel posStart posEnd label
+            let match (VarDecl _ l _) = l == label
+            case filter match fields of
+                x@(VarDecl _ _ t) : _ -> do
+                    Writer.tell [SourceCodeFragment src (VariableUse x)]
+                    return t
+                _ -> return AnyType
 
+        RecordLit (Map.toList -> l) -> handleRecordLike l
+
+        Record (Map.toList -> l) -> handleRecordLike l
```

Comment:
In which cases is it useful that we record a `RecordLiteral` for a `Record` _type_?

(I also wonder whether `RecordLiteral` should have a better name.)

Reply:
> (I also wonder whether `RecordLiteral` should have a better name.)

This is the problem. In the beginning, I used it only for record literals, and later I added support for `Record` using the same name.

I'll use `RecordFields`: generic and more appropriate with the `JtdInfo` type name you suggested